src/test_error/mod.rs

```diff
@@ -0,0 +1,54 @@
+// This file is part of rust-web/twig
+//
+// For the copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+
+/// Twig library for rust
+
+mod error_codes;
+use self::error_codes::{EngineResult, EngineErrorCode, RuntimeResult, RuntimeErrorCode};
+use error::api::ErrorCode;
+
+fn foo_engine_error() -> EngineResult<()> {
+    err!(EngineErrorCode::TemplateNotFound {
+        name: "hello_world.html.twig".to_string(),
+        search_paths: vec![]
+    })
+}
+
+fn foo_runtime_error() -> RuntimeResult<()> {
+    err!(RuntimeErrorCode::InvalidArgumentCount{
+        defined: 10,
+        given: 4
+    })
+}
+
+fn foo_chain_error() -> EngineResult<()> {
+    let _val: () = try_chain!(foo_runtime_error());
+
+    unreachable!()
+}
+
+#[test]
+fn test() {
+    // these testcases are of course highly unstable due to the error locations(!)
+    // -> just for demonstration right now:
+
+    let my_error = EngineErrorCode::RuntimeError
+        .at(location!())
+        .caused_by(foo_runtime_error().unwrap_err());
+
+    assert_eq!(my_error.to_string(),
+        "Some runtime error at src/test_error/mod.rs:38:12\n \
+        - caused by: Target requires 10 arguments, called with 4 at src/test_error/mod.rs:20:4\n");
```

Comment:
If the Error happened in a template, I would suggest something like that
```
    err!(SomeErrorCode::Foo{
        ...,
        cursor: Cursor // <-- information about current lexer position and template name
    })
```

This error would print like
```
Target requires 10 arguments, called with 4 for `index.twig.html` line 11 column 2 at src/test_error/mod.rs:20:4\n
```

You are right, that these error-chains might obscure the most valuable part from a users perspective. I have to admit, that I was designing this part from a developer perspective. This is why I wanted these pseudo-backtraces.

Maybe we can find a way to have two or even three modes to print errors. Currently there are already different stringify-methods for errors: Display::fmt + Debug::fmt + ErrorCode::description. Where Debug might give even _more_ information as my examples in some cases, because we can put really complex objects into these ErrorCodes. This might be useful. Display would also be useful in its current form. ErrorCode::description is only for compatibility with std::error::Error - and I am not yet 100% happy with it. Sadly the std Error-trait has a very restricting signature `description(&self) -> &str`, which makes it hard to add dynamic data without internal buffer strings. But it seems like this is how it was meant to be by the std Error-Trait - to use Display+Debug instead.

I would like to reserve the Debug::fmt of errors for local development - while Display::fmt should be something useful for production.

When we get bug reports I would like them to be as detailed as in the examples - with those pseudo-backtraces. But you are right - if the bug is in the template and not in the engine - it might be too much. So maybe we need some kind of verbosity-flag for displaying errors? I'm not sure, yet.